### PR TITLE
SessionToken: Adds optimization to remove session token for stored procedure, triggers, and UDFs

### PR DESCRIPTION
**Pull request by j82w**
# Pull Request Template

## Description

This fixes a bug where the session token was getting passed for stored procedures, triggers and user defined functions which caused 400 header to large exception. These types are not considered master resources because they execute on the data plane, but all the CRUD operations happen on master which does not require the session token.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] This change requires a documentation update

## Closing issues

To automatically close an issue: closes #IssueNumber

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 1]
```diff
@@ -189,14 +189,20 @@ public void Dispose()
 
         private void ApplySessionToken(DocumentServiceRequest request)
         {
-            if (request.Headers != null &&
-                !string.IsNullOrEmpty(request.Headers[HttpConstants.HttpHeaders.SessionToken]))
+            // Master resources don't require session token. Stored procedures, trigger, and user defined functions CRUD operations are done on
+            // master so they do not require the session token. 
+            if (ReplicatedResourceClient.IsMasterResource(request.ResourceType) ||
+                (request.ResourceType == ResourceType.StoredProcedure && request.OperationType != Documents.OperationType.ExecuteJavaScript) ||
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs:195]: nit> would it make sense to have internal static method called GatewayStoreModel.IsScriptRequest(request) to check this and make it easier to understand?

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 2]
```diff
@@ -189,14 +189,20 @@ public void Dispose()
 
         private void ApplySessionToken(DocumentServiceRequest request)
         {
-            if (request.Headers != null &&
-                !string.IsNullOrEmpty(request.Headers[HttpConstants.HttpHeaders.SessionToken]))
+            // Master resources don't require session token. Stored procedures, trigger, and user defined functions CRUD operations are done on
+            // master so they do not require the session token. 
+            if (ReplicatedResourceClient.IsMasterResource(request.ResourceType) ||
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs:194]: How about a new utility named IsMasterOperation(ResourceType, OperationType)?

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 3]
```diff
@@ -189,14 +189,20 @@ public void Dispose()
 
         private void ApplySessionToken(DocumentServiceRequest request)
         {
-            if (request.Headers != null &&
-                !string.IsNullOrEmpty(request.Headers[HttpConstants.HttpHeaders.SessionToken]))
+            // Master resources don't require session token. Stored procedures, trigger, and user defined functions CRUD operations are done on
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs:192]: It needs port to master as well right?

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs:192]: Already sent the v2 PR as well.

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs (hunk 2 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs:194]: Also might help UT?

File: Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs [hunk 4]
```diff
@@ -187,52 +191,76 @@ public void Dispose()
...
-            string sessionToken = this.sessionContainer.ResolveGlobalSessionToken(request);
+            string sessionToken = sessionContainer.ResolveGlobalSessionToken(request);
 
             if (!string.IsNullOrEmpty(sessionToken))
             {
                 request.Headers[HttpConstants.HttpHeaders.SessionToken] = sessionToken;
             }
         }
 
+        // DEVNOTE: This can be replace with ReplicatedResourceClient.IsMasterOperation on next Direct sync
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs:242]: Please add backlog to remove once direct is refreshed

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs:242]: https://github.com/Azure/azure-cosmos-dotnet-v3/issues/1771

**Pull request merged**